### PR TITLE
nydusd: graceful state machine thread shutdown

### PR DESCRIPTION
**Pull request by bergwolf**
When exiting, make sure we wait for the state machine thread to quit.

A few changes were applied:
1. exit_daemon() should call daemon.stop() so that it can result in
    different daemon action depending on the current state. Then in a normal
    workflow, nydusd would just umount and quit. But when in Interrupted,
    nydusd just dies directly without doing any umount etc.
2. Quit the state machine thread upon EXIT (TerminateFuseService) event.
3. Wait for the state machine thread together when waiting for the fuse
    service threads.
4. Nydusd should just die when getting EXIT message in INIT state.
5. Only send Stop event if state is not INTERRUPTED or STOPPED.
    Otherwise there is no state machine thread to handle it.
6. Init and Interrupted should always be able to stop as well.

### CODE REVIEW

**Comment by jiangliu** [comment]
Is this target 2.0? There will be heavy conflicts with fscache branch.

**Comment by bergwolf** [comment]
@jiangliu Yes, it fixes hot-upgrade and is targeting at 2.0 where hot-upgrade is broken with the introduction of mio.

File: src/bin/nydusd/api_server_glue.rs
Comment (imeoer) [inline src/bin/nydusd/api_server_glue.rs:229]: Is it required?

File: api/src/http.rs
Comment (bergwolf) [reply api/src/http.rs:430]: The PR handles EXIT_TOKEN. If poll fails, looks like we should also call exit_api_server()?

File: src/bin/nydusd/api_server_glue.rs
Comment (bergwolf) [reply src/bin/nydusd/api_server_glue.rs:229]: Yes, it is. We need to reply properly to API callers.

File: api/src/http.rs
Comment (bergwolf) [reply api/src/http.rs:430]: fixed

File: src/bin/nydusd/api_server_glue.rs
Comment (imeoer) [reply src/bin/nydusd/api_server_glue.rs:229]: The return value of `d.wait` seems not to be used, but that's okay.

**Review by imeoer — APPROVED** [review]

File: src/bin/nydusd/daemon.rs
Comment (liubogithub) [inline src/bin/nydusd/daemon.rs:241]: is it supposed to be "continue to serve"?

File: src/bin/nydusd/api_server_glue.rs
Comment (liubogithub) [inline src/bin/nydusd/api_server_glue.rs:224]: typo, "have been terminated"

File: src/bin/nydusd/fusedev.rs
Comment (liubogithub) [inline src/bin/nydusd/fusedev.rs:234]: This changes where threads.lock protects,  it'll unlock  `threads` after pop(), why is that?

**Comment by bergwolf** [comment]
Mark `wip` as I'm still testing it.

File: src/bin/nydusd/fusedev.rs
Comment (bergwolf) [reply src/bin/nydusd/fusedev.rs:234]: If we block on `handle.join` with the lock at hand, we cannot add more threads to the waiting list, which is necessary for the upgrade use cases, where we just start the state machine thread and then start more fuse service threads later on.

**Comment by jiangliu** [comment]
We need to merge this into fscache branch too once it gets merged:)

**Comment by bergwolf** [comment]
This is ready to be reviewed/merged.

File: src/bin/nydusd/fusedev.rs
Comment (liubogithub) [reply src/bin/nydusd/fusedev.rs:234]: Please correct me if I'm wrong.

In case of upgrade usecase, `d.wait()` is called by old binary's daemon right before old binary quits itself, so why would be more fuse service threads coming in to serve old binary?

With that being said, it looks like early unlock is fine because there is unlikely a race here.

Comment (bergwolf) [reply src/bin/nydusd/fusedev.rs:234]: It is not about the old binary's daemon. The new daemon will start and block on waiting for the state machine thread. And then the control path manager will send fd to the new daemon and ask it to take over the service. Then the new daemon will need to start fuse server threads and add them to the waiting list. That's why we need to drop the lock here.

**Pull request merged**
